add support for dependency management

jostillmanns opened the pull request:
please review this PR before the helmfs PR

jostillmanns committed:
add support for dependency management

johnholiver commented on cmd/compile/compile.go:
```diff
@@ -0,0 +1,69 @@
+package compile
```
We have a naming problem. Maybe layering.
Package and filename should be called `compiler`. Also, for some reason I don't understand why this is inside pligos/cmd/compiler. I think you might have just done it too fast. 

There's a pligos/pkg/compiler, is there a reason why this code isn't there? Also, there's Compile here that looks like a helper, and calls the real c.Compile(). :/

Just try to move the contents of this file to pligos/pkg/compiler and that will force you to put things potentially in the right place.

johnholiver commented on cmd/create/create.go:
```diff
@@ -0,0 +1,77 @@
+	return createTree(config.ConfigPath, config.ContextName, config.ChartPath)
+}
+
+func NewHelm(configPath, contextName string) (*helm.Helm, error) {
```
Seems that this method is in the wrong package too.
Besides, this method is the only one that calls helm.New().
Potentially, this should be in pligos/pkg/helm/helm.go

johnholiver commented on cmd/create/create.go:
```diff
@@ -0,0 +1,77 @@
+	return helm.New(context.Flavor, config.DeploymentConfig, config.ChartDependencies, context.Configs, context.Secrets, c), nil
+}
+
+func createTree(configPath, contextName, chartPath string) error {
```
This method could go to pkg/pligos, maybe?
If we need a struct to represent the chart tree, then we should have this abstraction explicitly declared, a go file to hold this, maybe a struct (needed?) and this method inside the file.

johnholiver commented on cmd/create/create.go:
```diff
@@ -0,0 +1,77 @@
+	"realcloud.tech/pligos/pkg/pligos"
+)
+
+type Config struct {
```
This type can go back to the cmd/create.go?

johnholiver reviewed (changes requested):
Feature looks good, architecture looks bad.
Moving methods and files around should suffice.
Besides, after reading this I imagine we could draw diagrams to explain pligos modules somehow.
I could start collaborating because I'm quite interested on it and I'll do the effort to understand the architecture thru code, but I doubt a random dev would be able to jump in. The diagram could be a good point to add to a Pligos presentation and documentation too. Do you think you could hand draw something and drop it on Slack? We could discuss on top of it, redraw and eventually transfer to a digital mean.

jostillmanns committed:
refactor to use the helm in-memory representation of a chart

also: to deal with chart dependencies, this adds a new command

  pligos helm update-dependencies

that needs to be called before create to make sure all dependencies
are present.

jostillmanns committed:
implement create file handing

jostillmanns committed:
make create a sub command of helm

jostillmanns committed:
fix leading `/` on config files

johnholiver commented on cmd/download.go:
```diff
@@ -0,0 +1,50 @@
+var downloadCmd = &cobra.Command{
+	Use:   "update-dependencies",
+	Short: "download helm dependencies",
+	Long:  `This will download the requirements defined by the requiremnets.yaml inside the pligos directory`,
```
`requiremnets` -> `requirements`

johnholiver commented on pkg/helm/dependency.go:
```diff
@@ -0,0 +1,75 @@
+
+// defaultKeyring returns the expanded path to the default keyring.
+func (d *DependencyManager) defaultKeyring() string {
+	return os.ExpandEnv("$HOME/.gnupg/pubring.gpg")
```
What is this?

johnholiver commented on pkg/pligos/config.go:
```diff
@@ -38,18 +48,24 @@ func FindContext(name string, contexts []Context) (Context, bool) {
 }
 
+type Dependency struct {
+	Pligos  string `yaml:"pligos" filepath:"resolve"`
```
What does Dependency.Pligos represent? It feels to me that it doesnt reflect a semantic meaning, otherwise I'd not have to ask.

jostillmanns replied:
path to pligos config

johnholiver commented on pkg/pligos/config.go:
```diff
@@ -28,6 +27,17 @@ type PligosConfig struct {
 	ChartDependencies []map[string]interface{} `yaml:"chartDependencies"`
 }
 
+type CreateConfig struct {
```
This new struct seems a lot like struct Context.

johnholiver commented on cmd/create/create.go:
```diff
@@ -0,0 +1,22 @@
+	ChartPath   string
+}
+
+func Create(config Config) error {
```
`cmd/compile/compile.go` and `cmd/create/create.go` look very different, and this fact introduces different code styles/standards already in this PR.

E.g.:

1. There's a pligos.CreateConfig but no pligos.CompileConfig, or any other pligos.CmdConfig.
2. Each command has a Config struct, but Create has struct Config and pligos.CreateConfig, why multiple configs? What's the meaning of this?
3. pligos.Compiler is a rich struct (has multiple attributes) and a method New(), helm.Creator has none, and you instantiate it without New() as in ln 20
4. The download command need to be observed as well, it follows the same problem, the code style doesn't seem to follow one logic. Maybe everything that was before downloader.Download should be encapsulated in a method. At least this seems to be the solution in compile and create.

johnholiver commented on pkg/pligos/make.go:
```diff
@@ -0,0 +1,106 @@
+	"realcloud.tech/pligos/pkg/pathutil"
+)
+
+func MakeCompiler(config PligosConfig, contextName string) (*compiler.Compiler, error) {
```
Why MakeCompiler receive a PligosConfig and MakeCreateConfig builds a PligosConfig from a pligosPath? This is inconsistent.

jostillmanns committed:
refactor everything to single command design

johnholiver commented on pkg/helmport/create.go:
```diff
@@ -0,0 +1,73 @@
+
+	var updatedTemplates []*chart.Template
+	for _, template := range p.Flavor.Templates {
+		newData := chartutil.Transform(string(template.Data), "<CHARTNAME>", p.Flavor.Metadata.Name)
```
Is this correct? Flavor.Metadata feels weird.
